File: packages/markdown-parser/src/core/parser.ts
// @fuyeor/markdown-parser/src/core/parser.ts
import { BlockState, InlineState } from './state';
import { linkify } from '@fuyeor/linkify';
import type {
  ASTNode,
  BlockRule,
  InlineRule,
  Linkifier,
  MarkdownPlugin,
  MarkdownParserOptions,
  ParserContext,
} from '#/types';
import { isSafeLinkUrl } from '#/core/url';

const LINKIFY_CANDIDATE_REGEX = /(?:https?:\/\/|[A-Za-z0-9-]+\.)/u;

/** Skip linkify scans for text without a possible URL candidate. */
function linkifyWithCandidateCheck(text: string): ReturnType<typeof linkify> {
  if (!LINKIFY_CANDIDATE_REGEX.test(text)) return [];
  return linkify(text);
}

export class MarkdownParser {
  #blockRuleMap = new Map<string, BlockRule[]>();
  #inlineRuleMap = new Map<string, InlineRule[]>();
  #idSequence = 0;
  #isPreflight = false;
  readonly #maxNestingDepth: number;
  readonly #linkifier: Linkifier;

  constructor(options: MarkdownParserOptions = {}) {
    const maxNestingDepth = options.maxNestingDepth ?? 64;
    if (!Number.isInteger(maxNestingDepth) || maxNestingDepth < 1)
      throw new RangeError('maxNestingDepth must be a positive integer');

    this.#maxNestingDepth = maxNestingDepth;
    this.#linkifier = options.linkifier ?? linkifyWithCandidateCheck;
  }

  // construct a context for recursive rule calls
  readonly #context: ParserContext = this.#createContext(0);

  #createContext(depth: number): ParserContext {
    return {
      parseInline: (content: string) =>
        this.#parseInline(new InlineState(content), depth + 1),
      parseBlocks: (content: string) =>
        this.#parseBlocks(new BlockState(content), depth + 1),
      createId: (prefix: string) =>
        this.#isPreflight
          ? `${prefix}-probe`
          : `${prefix}-${this.#idSequence++}`,
    };
  }

  // register block rule
  // By default, it inserts at the end, or before/after a specified rule.
  addBlockRule(rule: BlockRule): this {
    for (const marker of rule.markers) {
      const list = this.#blockRuleMap.get(marker) || [];
      list.push(rule);
      this.#blockRuleMap.set(marker, list);
    }
    return this;
  }

  // register inline rule
  addInlineRule(rule: InlineRule): this {
    // register the rule under its markers for quick lookup during parsing
    for (const marker of rule.markers) {
      const rulesForMarker = this.#inlineRuleMap.get(marker) || [];
      // ensure the rule is in the list for this marker
      rulesForMarker.unshift(rule);
      this.#inlineRuleMap.set(marker, rulesForMarker);
    }

    return this;
  }

  use(plugin: MarkdownPlugin): this {
    plugin(this);
    return this;
  }

  build(): (content: string) => ASTNode[] {
    return (content: string) => {
      this.#idSequence = 0;
      const state = new BlockState(content);
      return this.#parseBlocks(state);
    };
  }

  #parseBlocks(state: BlockState, depth = 0): ASTNode[] {
    if (depth > this.#maxNestingDepth) {
      return [
        {
          type: 'paragraph',
          children: [
            {
              type: 'text',
              content: state.remainingLines.join('\n'),
            },
          ],
        },
      ];
    }

    const context = depth === 0 ? this.#context : this.#createContext(depth);
    const nodes: ASTNode[] = [];
    while (state.lineIndex < state.lineCount) {
      const line = state.currentLine;

      // skip purely empty line
      if (!line || line.trim() === '') {
        state.advance(1);
        continue;
      }

      // find the first non-whitespace character
      const trimmed = line.trimStart();
      const firstChar = trimmed[0];

      // get candidate rules based on the first character
      const rules = this.#blockRuleMap.get(firstChar);

      let matched = false;

      if (rules) {
        for (const rule of rules) {
          const result = rule.parse(state, context);
          if (result) {
            nodes.push(result.node);
            state.advance(result.consumedLines);
            matched = true;
            break;
          }
        }
      }

      // continuous text merged into a single paragraph
      if (!matched) {
        const paragraphLines: string[] = [];

        while (state.lineIndex < state.lineCount) {
          const line = state.currentLine;
          if (!line || line.trim() === '') break;

          // fast-path character preflight
          const firstChar = line.trimStart().charCodeAt(0);
          const mayInterrupt =
            // # for title
            firstChar === 35 ||
            // * for bold and italic
            firstChar === 42 ||
            // _ for underline
            firstChar === 95 ||
            // +
            firstChar === 43 ||
            // - for bulletin list and strike
            firstChar === 45 ||
            // 0-9 for ordered list
            (firstChar >= 48 && firstChar <= 57) ||
            // > for inline quote
            firstChar === 62 ||
            // ` for code block and inline code
            firstChar === 96 ||
            // ~ equals `
            firstChar === 126 ||
            // $ for block math
            firstChar === 36;

          if (mayInterrupt) {
            const rules = this.#blockRuleMap.get(
              String.fromCharCode(firstChar),
            );
            let isInterrupted = false;

            if (rules) {
              for (const rule of rules) {
                if (
                  [
                    'heading',
                    'hr',
                    'blockquote',
                    'list',
                    'code_block',
                    'math_block',
                    'ffm_blocks',
                  ].includes(rule.name)
                ) {
                  this.#isPreflight = true;
                  try {
                    if (rule.parse(state, context) !== null) {
                      isInterrupted = true;
                      break;
                    }
                  } finally {
                    this.#isPreflight = false;
                  }
                }
              }
            }

            if (isInterrupted && paragraphLines.length > 0) break;
          }

          paragraphLines.push(line);
          state.advance(1);
        }

        if (paragraphLines.length > 0) {
          nodes.push({
            type: 'paragraph',
            children: this.#parseInline(
              new InlineState(paragraphLines.join('\n')),
              depth,
            ),
          });
        }
      }
    }
    return nodes;
  }

  #parseInline(state: InlineState, depth = 0): ASTNode[] {
    if (depth > this.#maxNestingDepth) {
      return [{ type: 'text', content: state.content.slice(state.pos) }];
    }

    const context = depth === 0 ? this.#context : this.#createContext(depth);
    const nodes: ASTNode[] = [];

    // record the start position of plain text
    let textStart = state.pos;

    // linkify
    const flushText = (endPos: number) => {
      if (endPos > textStart) {
        const textBuffer = state.content.slice(textStart, endPos);
        let lastIdx = 0;

        for (const match of this.#linkifier(textBuffer)) {
          if (isSafeLinkUrl(match.url)) {
            if (match.index > lastIdx) {
              nodes.push({
                type: 'text',
                content: textBuffer.slice(lastIdx, match.index),
              });
            }
            nodes.push({
              type: 'link',
              url: match.url,
              children: [{ type: 'text', content: match.text }],
            });
            lastIdx = match.lastIndex;
          }
        }

        if (lastIdx < textBuffer.length) {
          nodes.push({ type: 'text', content: textBuffer.slice(lastIdx) });
        }
      }
    };

    while (state.pos < state.length) {
      const char = state.currentChar;

      // find candidate rules based on the current character
      const rules = char ? this.#inlineRuleMap.get(char) : undefined;

      if (rules) {
        let matched = false;

        // only check rules that are registered for this marker character
        for (const rule of rules) {
          const result = rule.parse(state, context);
          if (result) {
            flushText(state.pos);
            nodes.push(result.node);
            state.advance(result.consumedChars);
            textStart = state.pos;
            matched = true;
            break;
          }
        }

        // skip the normal character advancement if any rule matched
        if (matched) continue;
      }

      // normal characters without matching rules, just advance
      if (char === '\n') {
        flushText(state.pos);
        nodes.push({ type: 'hardbreak' });
        state.advance(1);
        textStart = state.pos;
        continue;
      }

      // just advance the pointer for normal characters
      state.advance(1);
    }

    flushText(state.pos);
    return nodes;
  }
}

File: packages/markdown-parser-rust/src/parser.rs
// packages/markdown-parser-rust/src/parser.rs
use std::cell::RefCell;
use std::collections::HashMap;

use crate::ast::{AstNode, NodeType};
use crate::linkify::{LinkMatch, linkify};
use crate::rules::{BlockRule, FfmBlockRule, InlineRule};
use crate::safety::is_safe_link_url;
use crate::state::{BlockState, InlineState};

const LINKIFY_CANDIDATE_MARKERS: &[char] = &['.'];
const PRELIGHT_BLOCK_RULES: &[&str] = &[
    "heading",
    "hr",
    "blockquote",
    "list",
    "code_block",
    "ffm_blocks",
];

/// A linkifier callback used by the parser's inline text projection.
pub type Linkifier = fn(&str) -> Vec<LinkMatch>;

/// Parser options matching the TypeScript parser constructor.
#[derive(Clone, Copy)]
pub struct ParserOptions {
    /// Maximum recursive block/inline parsing depth.
    pub max_nesting_depth: usize,
    /// Callback used to discover links in plain text.
    pub linkifier: Linkifier,
}

impl Default for ParserOptions {
    fn default() -> Self {
        Self {
            max_nesting_depth: 64,
            linkifier: linkify,
        }
    }
}

/// Errors returned when parser configuration is invalid.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ParserError {
    /// The configured maximum nesting depth is zero.
    InvalidMaxNestingDepth,
}

impl std::fmt::Display for ParserError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::InvalidMaxNestingDepth => {
                formatter.write_str("max_nesting_depth must be a positive integer")
            }
        }
    }
}

impl std::error::Error for ParserError {}

/// Recursive parser context passed to custom block and inline rules.
pub struct ParserContext<'a> {
    pub(crate) parser: &'a MarkdownParser,
    pub(crate) depth: usize,
    runtime: &'a RefCell<ParseRuntime>,
}

impl ParserContext<'_> {
    /// Recursively parses inline content using the current parser configuration.
    pub fn parse_inline(&self, content: &str) -> Vec<AstNode> {
        self.parser
            .parse_inline_inner(content, self.depth + 1, self.runtime)
    }

    /// Recursively parses block content using the current parser configuration.
    pub fn parse_blocks(&self, content: &str) -> Vec<AstNode> {
        self.parser
            .parse_blocks_inner(content, self.depth + 1, self.runtime)
    }

    /// Creates an ID with the probe suffix during paragraph preflight.
    pub fn create_id(&self, prefix: &str) -> String {
        if self.runtime.borrow().is_preflight {
            format!("{prefix}-probe")
        } else {
            let mut runtime = self.runtime.borrow_mut();
            let sequence = runtime.id_sequence;
            runtime.id_sequence += 1;
            format!("{prefix}-{sequence}")
        }
    }
}

/// State isolated to one parser invocation.
#[derive(Default)]
struct ParseRuntime {
    id_sequence: usize,
    is_preflight: bool,
}

/// A lightweight, extensible Markdown and FFM parser.
pub struct MarkdownParser {
    block_rules: Vec<Box<dyn BlockRule>>,
    inline_rules: Vec<Box<dyn InlineRule>>,
    block_rule_map: HashMap<char, Vec<usize>>,
    inline_rule_map: HashMap<char, Vec<usize>>,
    max_nesting_depth: usize,
    linkifier: Linkifier,
}

impl MarkdownParser {
    /// Creates a parser and fails fast if its options are invalid.
    pub fn new(options: ParserOptions) -> Self {
        Self::try_new(options).expect("invalid MarkdownParser options")
    }

    /// Creates a parser without panicking on invalid configuration.
    pub fn try_new(options: ParserOptions) -> Result<Self, ParserError> {
        if options.max_nesting_depth == 0 {
            return Err(ParserError::InvalidMaxNestingDepth);
        }
        Ok(Self {
            block_rules: Vec::new(),
            inline_rules: Vec::new(),
            block_rule_map: HashMap::new(),
            inline_rule_map: HashMap::new(),
            max_nesting_depth: options.max_nesting_depth,
            linkifier: options.linkifier,
        })
    }

    /// Registers a block rule and appends it to the matching marker lists.
    pub fn add_block_rule<R>(&mut self, rule: R) -> &mut Self
    where
        R: BlockRule + 'static,
    {
        let rule_index = self.block_rules.len();
        let markers = rule.markers();
        self.block_rules.push(Box::new(rule));
        for marker in markers {
            self.block_rule_map
                .entry(*marker)
                .or_default()
                .push(rule_index);
        }
        self
    }

    /// Registers an inline rule at the front of each matching marker list.
    pub fn add_inline_rule<R>(&mut self, rule: R) -> &mut Self
    where
        R: InlineRule + 'static,
    {
        let rule_index = self.inline_rules.len();
        let markers = rule.markers();
        self.inline_rules.push(Box::new(rule));
        for marker in markers {
            self.inline_rule_map
                .entry(*marker)
                .or_default()
                .insert(0, rule_index);
        }
        self
    }

    /// Installs a plugin that can register rules or otherwise configure the parser.
    pub fn use_plugin(&mut self, plugin: fn(&mut MarkdownParser)) -> &mut Self {
        plugin(self);
        self
    }

    /// Parses a complete document and resets generated FFM IDs for this invocation.
    pub fn parse(&self, content: &str) -> Vec<AstNode> {
        let runtime = RefCell::new(ParseRuntime::default());
        self.parse_blocks_inner(content, 0, &runtime)
    }

    /// Parses the input with the standard rule set.
    pub fn create_standard(options: ParserOptions) -> Self {
        let mut parser = Self::new(options);
        parser.add_block_rule(crate::rules::CodeBlockRule);
        parser.add_block_rule(crate::rules::ListRule);
        parser.add_block_rule(crate::rules::HeadingRule);
        parser.add_block_rule(crate::rules::TableRule);
        parser.add_block_rule(crate::rules::HrRule);
        parser.add_block_rule(crate::rules::BlockquoteRule);
        parser.add_inline_rule(crate::rules::HardBreakRule);
        parser.add_inline_rule(crate::rules::InlineCodeRule);
        parser.add_inline_rule(crate::rules::LinkRule);
        parser.add_inline_rule(crate::rules::BoldRule);
        parser.add_inline_rule(crate::rules::UnderlineRule);
        parser.add_inline_rule(crate::rules::ItalicRule);
        parser.add_inline_rule(crate::rules::StrikeRule);
        parser
    }

    /// Parses the input with the standard rule set plus FFM fenced blocks.
    pub fn create_ffm(options: ParserOptions) -> Self {
        let mut parser = Self::new(options);
        parser.add_block_rule(FfmBlockRule);
        parser.add_block_rule(crate::rules::CodeBlockRule);
        parser.add_block_rule(crate::rules::ListRule);
        parser.add_block_rule(crate::rules::HeadingRule);
        parser.add_block_rule(crate::rules::TableRule);
        parser.add_block_rule(crate::rules::HrRule);
        parser.add_block_rule(crate::rules::BlockquoteRule);
        parser.add_inline_rule(crate::rules::HardBreakRule);
        parser.add_inline_rule(crate::rules::InlineCodeRule);
        parser.add_inline_rule(crate::rules::LinkRule);
        parser.add_inline_rule(crate::rules::BoldRule);
        parser.add_inline_rule(crate::rules::UnderlineRule);
        parser.add_inline_rule(crate::rules::ItalicRule);
        parser.add_inline_rule(crate::rules::StrikeRule);
        parser
    }

    /// Parses block content at a specific recursion depth.
    fn parse_blocks_inner(
        &self,
        content: &str,
        depth: usize,
        runtime: &RefCell<ParseRuntime>,
    ) -> Vec<AstNode> {
        let mut state = BlockState::new(content);
        self.parse_blocks_state(&mut state, depth, runtime)
    }

    /// Runs block rules and merges unmatched consecutive lines into paragraphs.
    fn parse_blocks_state(
        &self,
        state: &mut BlockState,
        depth: usize,
        runtime: &RefCell<ParseRuntime>,
    ) -> Vec<AstNode> {
        if depth > self.max_nesting_depth {
            return vec![AstNode::with_children(
                NodeType::Paragraph,
                vec![AstNode::text(state.remaining_lines().join("\n"))],
            )];
        }

        let context = ParserContext {
            parser: self,
            depth,
            runtime,
        };
        let mut nodes = Vec::new();
        while state.line_index < state.lines.len() {
            let Some(line) = state.current_line() else {
                break;
            };
            if line.is_empty() || line.trim().is_empty() {
                state.advance(1);
                continue;
            }

            let first_char = line.trim_start().chars().next();
            let mut matched = false;
            if let Some(first_char) = first_char
                && let Some(rule_indices) = self.block_rule_map.get(&first_char)
            {
                for rule_index in rule_indices {
                    if let Some(result) = self.block_rules[*rule_index].parse(state, &context) {
                        nodes.push(result.node);
                        state.advance(result.consumed_lines);
                        matched = true;
                        break;
                    }
                }
            }

            if matched {
                continue;
            }

            let mut paragraph_lines = Vec::new();
            while state.line_index < state.lines.len() {
                let Some(current_line) = state.current_line() else {
                    break;
                };
                if current_line.is_empty() || current_line.trim().is_empty() {
                    break;
                }
                let first = current_line.trim_start().chars().next();
                let may_interrupt = first.is_some_and(|character| {
                    matches!(
                        character,
                        '#' | '*' | '_' | '+' | '-' | '0'..='9' | '>' | '`' | '~'
                    )
                });
                if may_interrupt {
                    let mut is_interrupted = false;
                    if let Some(first) = first
                        && let Some(rule_indices) = self.block_rule_map.get(&first)
                    {
                        for rule_index in rule_indices {
                            let rule = &self.block_rules[*rule_index];
                            if PRELIGHT_BLOCK_RULES.contains(&rule.name()) {
                                runtime.borrow_mut().is_preflight = true;
                                let result = rule.parse(state, &context).is_some();
                                runtime.borrow_mut().is_preflight = false;
                                if result {
                                    is_interrupted = true;
                                    break;
                                }
                            }
                        }
                    }
                    if is_interrupted && !paragraph_lines.is_empty() {
                        break;
                    }
                }
                paragraph_lines.push(current_line.to_owned());
                state.advance(1);
            }

            if !paragraph_lines.is_empty() {
                let inline_content = paragraph_lines.join("\n");
                nodes.push(AstNode::with_children(
                    NodeType::Paragraph,
                    self.parse_inline_inner(&inline_content, depth, runtime),
                ));
            }
        }
        nodes
    }

    /// Parses inline content at a specific recursion depth.
    fn parse_inline_inner(
        &self,
        content: &str,
        depth: usize,
        runtime: &RefCell<ParseRuntime>,
    ) -> Vec<AstNode> {
        let mut state = InlineState::new(content);
        self.parse_inline_state(&mut state, depth, runtime)
    }

    /// Runs marker-selected inline rules and linkifies the remaining text.
    fn parse_inline_state(
        &self,
        state: &mut InlineState,
        depth: usize,
        runtime: &RefCell<ParseRuntime>,
    ) -> Vec<AstNode> {
        if depth > self.max_nesting_depth {
            return vec![AstNode::text(&state.content[state.pos..])];
        }

        let context = ParserContext {
            parser: self,
            depth,
            runtime,
        };
        let mut nodes = Vec::new();
        let mut text_start = state.pos;
        while state.pos < state.content.len() {
            let Some(character) = state.current_char() else {
                break;
            };
            if let Some(rule_indices) = self.inline_rule_map.get(&character).cloned() {
                let mut matched = false;
                for rule_index in rule_indices {
                    if let Some(result) = self.inline_rules[rule_index].parse(state, &context) {
                        self.flush_text(
                            &mut nodes,
                            &state.content[text_start..state.pos],
                            text_start,
                        );
                        nodes.push(result.node);
                        state.advance(result.consumed_bytes);
                        text_start = state.pos;
                        matched = true;
                        break;
                    }
                }
                if matched {
                    continue;
                }
            }

            if character == '\n' {
                self.flush_text(
                    &mut nodes,
                    &state.content[text_start..state.pos],
                    text_start,
                );
                nodes.push(AstNode::new(NodeType::Hardbreak));
                state.advance(1);
                text_start = state.pos;
                continue;
            }
            state.advance(character.len_utf8());
        }
        self.flush_text(&mut nodes, &state.content[text_start..], text_start);
        nodes
    }

    /// Converts a plain-text range into safe links and text nodes.
    fn flush_text(&self, nodes: &mut Vec<AstNode>, text: &str, _offset: usize) {
        if text.is_empty() || !has_linkify_candidate(text) {
            if !text.is_empty() {
                nodes.push(AstNode::text(text));
            }
            return;
        }

        let mut last_index = 0;
        for matched in (self.linkifier)(text) {
            if !is_safe_link_url(&matched.url) {
                continue;
            }
            if matched.index > last_index {
                nodes.push(AstNode::text(&text[last_index..matched.index]));
            }
            let mut link = AstNode::new(NodeType::Link);
            link.url = Some(matched.url);
            link.children = vec![AstNode::text(matched.text)];
            nodes.push(link);
            last_index = matched.last_index;
        }
        if last_index < text.len() {
            nodes.push(AstNode::text(&text[last_index..]));
        }
    }
}

/// Avoids the linkifier call for text that cannot contain an URL candidate.
fn has_linkify_candidate(text: &str) -> bool {
    text.contains("://")
        || text
            .chars()
            .any(|character| LINKIFY_CANDIDATE_MARKERS.contains(&character))
}
